Source/Global/global.cpp

```diff
@@ -13,10 +13,129 @@
 
 using namespace xbox::httpclient;
 
-static std::shared_ptr<http_singleton> g_httpSingleton_atomicReadsOnly;
-
 NAMESPACE_XBOX_HTTP_CLIENT_BEGIN
 
+enum class singleton_access_mode
+{
+    get,
+    set
+};
+
+std::shared_ptr<http_singleton> singleton_access(
+    _In_ singleton_access_mode mode,
+    _In_opt_ std::shared_ptr<http_singleton> singleton = nullptr
+) noexcept
+{
+    static std::mutex s_mutex;
+    static std::shared_ptr<http_singleton> s_singleton{ nullptr };
+
+    std::lock_guard<std::mutex> lock{ s_mutex };
+    switch (mode)
+    {
+    case singleton_access_mode::get:
+    {
+        assert(!singleton);
+        return s_singleton;
+    }
+    case singleton_access_mode::set:
+    {
+        std::swap(s_singleton, singleton);
+        return singleton;
+    }
+    default: 
+    {
+        assert(false);
+        return nullptr;
+    }
+    }
+}
+
+HRESULT http_singleton::create(
+    _In_ HCInitArgs* args
+) noexcept
+{
+    if (singleton_access(singleton_access_mode::get))
+    {
+        return E_HC_ALREADY_INITIALISED;
+    }
+
+    PerformEnv performEnv;
+    auto hr = Internal_InitializeHttpPlatform(args, performEnv);
+
+    if (SUCCEEDED(hr))
+    {
+        auto rawSingleton = new (http_memory::mem_alloc(sizeof(http_singleton))) http_singleton(
+            GetUserHttpPerformHandler(),
+#if !HC_NOWEBSOCKETS
+            GetUserWebSocketPerformHandlers(),
+#endif
+            std::move(performEnv)
+            );
+
+        auto singleton = std::shared_ptr<http_singleton>(rawSingleton, http_alloc_deleter<http_singleton>());
+
+        // Store self reference to be remove on http_singleton::cleanup_async
+        singleton->m_self = singleton;
+
+        singleton = singleton_access(singleton_access_mode::set, singleton);
+    }
+
+    return S_OK;
+}
+
+HRESULT http_singleton::cleanup_async(
+    _In_ XAsyncBlock* async
+) noexcept
+{
+    auto singleton{ singleton_access(singleton_access_mode::set, nullptr) };
+    if (!singleton)
+    {
+        E_HC_NOT_INITIALISED;
+    }
+
+    return XAsyncBegin(
+        async,
+        singleton.get(),
+        reinterpret_cast<void*>(cleanup_async),
+        __FUNCTION__,
+        [](XAsyncOp op, const XAsyncProviderData* data)
+    {
+        switch (op)
+        {
+        case XAsyncOp::Begin:
+        {
+            return XAsyncSchedule(data->async, 0);
+        }
+        case XAsyncOp::DoWork:
+        {
+            auto singleton{ static_cast<http_singleton*>(data->context)->shared_from_this() };
+            singleton->m_self.reset();
+
+            shared_ptr_cache::cleanup(singleton);
+
+            // Wait for all other references to the singleton to go away
+            // Note that the use count check here is only valid because we never create
+            // a weak_ptr to the singleton. If we did that could cause the use count
+            // to increase even though we are the only strong reference
+            while (singleton.use_count() > 1)
```

Comment:
uh... I don't know why I didn't notice this back when we did the Xsapi change, but this could cause deadlocks unless we are 100% sure no other work item produced by libhttpclient will ever schedule itself twice

I think the simplest fix would be to, instead of doing a busy loop, to call `XAsyncSchedule` again with a small delay, therefore freeing the queue to process other work

Reply:
If you take this suggestion, you'll have to move the `m_self.reset` until after all the uses are gone